### PR TITLE
Barel/functions

### PR DESCRIPTION
**Pull request by BarelUklis**
added endpoint to add coins to the user coins arrray

### CODE REVIEW

File: functions/src/socialAPIs/reddit/reddit-controller.ts [hunk 1]
```diff
@@ -14,7 +14,19 @@ export const getRedditPostsApi = async (data: string) => {
...
         
-        const redditData = await getRedditData(socialInfoFromCoinGecko.reddit);
+        let redditData = await getRedditData(socialInfoFromCoinGecko.reddit);
+
+        socialInfoFromCoinGecko.name && await addSocialDataToDB(socialInfoFromCoinGecko.name, 'reddit', redditData.slice(0, 5));
+
+        if(data.order === 'newest') {
+            // sort by date and set the length to 3
+            redditData = redditData.sort((a: any, b: any) => (a.date) > (b.date) ? -1 : 1).slice(0, 3);
+        }
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/reddit/reddit-controller.ts:24]: else or switch

File: functions/src/socialAPIs/twitter/getUserTweets.ts [hunk 2]
```diff
@@ -1,14 +1,15 @@
 /* tslint:disable */
 import axios from "axios";
 import {IUserTweet} from "../../interfaces/user-tweets-interface";
+import { addSocialDataToDB } from "../utils";
 import getTweetById from "./getTweetById";
 
-const getUserTweets = async (userId: string, order: string) => {
+const getUserTweets = async (userId: string, order: string, nameFromCoinGeck: string | undefined) => {
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/twitter/getUserTweets.ts:7]: userId: string, order: string, nameFromCoinGeck: string = ''

File: functions/src/socialAPIs/telegram/telegram-controller.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import getCoinSocialMedias from "../coins/coin-gecko";
 
-export const getTelegramPostsApi = async (data: string) => {
-    const socialInfoFromCoinGecko = await getCoinSocialMedias(data);
+export const getTelegramPostsApi = async (data: {name: string, order: string}) => {
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/telegram/telegram-controller.ts:3]: {name: string, order: string} extract to interface

File: functions/src/socialAPIs/twitter/twitter-controller.ts [hunk 4]
```diff
@@ -30,18 +29,27 @@ export const getTwitterPostsApi = async (data: {name: string, order: string}) =>
...
-        const socialInfoFromCoinGecko = await getCoinSocialMedias(data);
+    export const getTwitterPostsByQueryApi = async (data: {name: string, order: string}) => {
+        const socialInfoFromCoinGecko = await getCoinSocialMedias(data.name);
     
         setBasicInfo(socialInfoFromCoinGecko)
     
         if (socialInfoFromCoinGecko.twitter && socialInfoFromCoinGecko.name) {
             socailObject.twitter = {
                 twitterFollowers: 0,
                 posts: []
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/twitter/twitter-controller.ts:40]: add type

File: functions/src/socialAPIs/twitter/twitter-controller.ts [hunk 5]
```diff
@@ -30,18 +29,27 @@ export const getTwitterPostsApi = async (data: {name: string, order: string}) =>
...
-            };
+            };            
     
-            const twitterData = await getTweetByQuery(socialInfoFromCoinGecko.name)
+            let twitterData = await getTweetByQuery(socialInfoFromCoinGecko.name);    
+            await addSocialDataToDB(socialInfoFromCoinGecko.name, 'twitter-by-query', twitterData.slice(0, 5));
+        
+            if(data.order === 'newest') {
+                twitterData = twitterData.slice(0, 3);
+            }
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/twitter/twitter-controller.ts:48]: add else if

File: functions/src/socialAPIs/twitter/getUserTweets.ts [hunk 6]
```diff
@@ -1,14 +1,15 @@
...
     const bearerToken = 'AAAAAAAAAAAAAAAAAAAAAI4QhQEAAAAAChhSJ6cnBzgigR2P59G1Qov71G4%3DGTIkGH08uPADvtiabLdVuVWFvs9fTHp05MEtgvHfvD11STLE4p'
     const url = `https://api.twitter.com/2/users/${userId}/tweets`;
 
     let userTweets: IUserTweet[] = [];
-    let resp;
+    let resp: any;
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/twitter/getUserTweets.ts:12]: type

File: functions/src/socialAPIs/utils.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+};
+
+export const addSocialDataToDB = async (coin: string, platform: string, data: any) => {  
+    const coinFromDb = await admin.firestore()
+        .collection('coins')
+        .doc(coin)
+        .collection(platform)
+        // get the lasted doc
+        .orderBy('createdAt', 'desc')
+        .limit(1)
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/utils.ts:29]: check if you need this

File: functions/src/socialAPIs/utils.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        .collection(platform)
+        .doc(`${lastDoc ? lastDoc.data().version + 1 : 1}`)
+        .set({
+          data,
+          version: lastDoc ? lastDoc.data().version + 1 : 1,
+          createdAt: admin.firestore.Timestamp.now()
+        })
+    }
+
+    if(lastDoc && lastDoc.data().data[0].id === data[0].id) {
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/utils.ts:48]: check if need else instead of if

File: functions/src/socialAPIs/utils.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        .doc(coin)
+        .collection(platform)
+        // get the lasted doc
+        .orderBy('createdAt', 'desc')
+        .limit(1)
+        .get();        
+
+    const lastDoc = coinFromDb.docs[0];
+
+    if(!lastDoc || (lastDoc && lastDoc.data().data[0].id !== data[0].id)) {
```
Comment (YuvalKrispin) [inline functions/src/socialAPIs/utils.ts:34]: `if(!lastDoc || lastDoc.data().data[0].id !== data[0].id) {`

File: functions/src/usersAPIs/users-controller.ts [hunk 10]
```diff
@@ -84,5 +85,108 @@ export const loginWithCredentialsApi = async (data: IUser) => {
...
+        // remove the tag
+        const newTags = coins[data.symbol].tags.filter((tag: string) => tag !== data.tag);
+        await admin.firestore()
+          .collection('users')
+          .doc(userId)
+          .update({
+            [`coins.${data.symbol}.tags`]: newTags,
+          });          
+        // if there are no tags at all, remove the coin from the coins arra        
+        if (coins[data.symbol].tags.length === 1 && coins[data.symbol].tags[0] === data.tag) {
```
Comment (YuvalKrispin) [inline functions/src/usersAPIs/users-controller.ts:115]: `coins[data.symbol].tags?.length === 1 && coins[data.symbol].tags[0] === data.tag`

File: functions/src/usersAPIs/users-controller.ts [hunk 11]
```diff
@@ -84,5 +85,108 @@ export const loginWithCredentialsApi = async (data: IUser) => {
...
+          await admin.firestore()
+            .collection('users')
+            .doc(userId)
+            .update({
+              coins,
+            });
+        }
+        return 'tag removed';
+      }
+      if (!tagExists && coinExists) {
```
Comment (YuvalKrispin) [inline functions/src/usersAPIs/users-controller.ts:148]:
```
      if (!tagExists) {     
   if(coinExists){

        // add the tag
        await admin.firestore()
          .collection('users')
          .doc(userId)
          .update({
            [`coins.${data.symbol}.tags`]: admin.firestore.FieldValue.arrayUnion(data.tag),
          });
          return 'coin added';
      } 
} else {
// add the coin
        await admin.firestore()
          .collection('users')
          .doc(userId)
          .update({
            [`coins.${data.symbol}`]: {
              tags: [data.tag],
              createdAt: admin.firestore.FieldValue.serverTimestamp(),
              //versions: data.versions,
            },
          });
        return 'coin added';
}
}
```

File: functions/src/usersAPIs/users-controller.ts [hunk 12]
```diff
@@ -84,5 +85,108 @@ export const loginWithCredentialsApi = async (data: IUser) => {
...
+        .update({
+          [`coins.${data.symbol}`]: {
+            tags: [data.tag],
+            createdAt: admin.firestore.FieldValue.serverTimestamp(),
+            //versions: data.versions,
+          },
+        });
+      return 'Coin added to user';
+    }
+  }
```
Comment (YuvalKrispin) [inline functions/src/usersAPIs/users-controller.ts:185]: else {

File: src/store/user-store/index.ts [hunk 13]
```diff
@@ -8,7 +8,8 @@ class UserStore {
 
   userData: IUserData = {
     email: '',
-    password: ''
+    password: '',
+    coins: []
```
Comment (YuvalKrispin) [inline src/store/user-store/index.ts:12]: type

File: src/store/user-store/index.ts [hunk 14]
```diff
@@ -66,6 +67,20 @@ class UserStore {
   setPassword(userPassword: string) {
     this.userData.password = userPassword;
   }
+
+  addCoin(coin: any) {
```
Comment (YuvalKrispin) [inline src/store/user-store/index.ts:71]: type

File: src/store/user-store/index.ts [hunk 15]
```diff
@@ -66,6 +67,20 @@ class UserStore {
...
+    if(this.userData.coins) {
+      this.userData.coins.push(coin);
+      functionsCall('addCoinToUserApi')(coin).then((result: any) => {
```
Comment (YuvalKrispin) [inline src/store/user-store/index.ts:74]: type

File: src/store/user-store/index.ts [hunk 16]
```diff
@@ -66,6 +67,20 @@ class UserStore {
...
+        console.log(result);
+      });
+    } else {
+      this.userData.coins = [coin];
+      functionsCall('addCoinToUserApi')(coin).then((result: any) => {
```
Comment (YuvalKrispin) [inline src/store/user-store/index.ts:79]: type

File: src/store/user-store/index.ts [hunk 17]
```diff
@@ -66,6 +67,20 @@ class UserStore {
   setPassword(userPassword: string) {
     this.userData.password = userPassword;
   }
+
+  addCoin(coin: any) {
+    if(this.userData.coins) {
+      this.userData.coins.push(coin);
+      functionsCall('addCoinToUserApi')(coin).then((result: any) => {
+        console.log(result);
```
Comment (YuvalKrispin) [inline src/store/user-store/index.ts:75]: remove console logs!!!!

File: functions/src/socialAPIs/utils.ts (hunk 7 above)
Comment (BarelUklis) [reply functions/src/socialAPIs/utils.ts:29]: need to add 1 other wise its send all the data

File: src/store/user-store/index.ts (hunk 15 above)
Comment (BarelUklis) [reply src/store/user-store/index.ts:74]: need a function to push coins to the user coins array?

Comment (YuvalKrispin) [reply src/store/user-store/index.ts]: ???

**Pull request merged**